kernel/george.c

```diff
@@ -58,23 +58,37 @@ typedef struct {
   uint32_t unused;
 } multiboot_module;
 
-void kernel_main (uint32_t mboot_struct_addr) {
+extern int _end;
+
+void kernel_main (uint32_t mboot_struct_addr, uint32_t mboot_magic) {
+  if (mboot_magic != 0x2BADB002) {
+    panic("Bad multiboot magic!");
+  }
   uint32_t *mboot_struct =
       (uint32_t *)phys_to_virt((uint64_t)mboot_struct_addr);
-  memtop = ((uint64_t)mboot_struct[2]) * 1024;
-  memtop += 0x100000;
+  uint64_t memtop = ((uint64_t)mboot_struct[2]) * 1024 + 0x100000;
   clear_screen();
   puts("Welcome to GeorgeOS, Multiboot Edition!\r\n");
-  initialize_allocator();
+
+  /* Find the end of the last multiboot module */
+  uint64_t lowest_safe_physical_address = (uint64_t)&_end - 0xFFFFFFFF80000000;
```

Comment:
What's more, there are two similar looking constants like it... it's definitely sad.